flyway-core/src/main/java/org/flywaydb/core/internal/scanner/classpath/OsgiClassPathLocationScanner.java

```diff
@@ -58,21 +56,17 @@
         return resourceNames;
     }
 
+    // Search through all bundles until we find one with the right host
     private Bundle getTargetBundleOrCurrent(Bundle currentBundle, URL locationUrl) {
-        try {
-            Bundle targetBundle = currentBundle.getBundleContext().getBundle(getBundleId(locationUrl.getHost()));
-            return targetBundle != null ? targetBundle : currentBundle;
-        } catch (Exception e) {
-            return currentBundle;
-        }
-    }
-
-    private long getBundleId(String host) {
-        final Matcher matcher = bundleIdPattern.matcher(host);
-        if (matcher.find()) {
-            return Double.valueOf(matcher.group()).longValue();
-        }
-        throw new IllegalArgumentException("There's no bundleId in passed URL");
+        final String path = locationUrl.getPath();
+        final String host = locationUrl.getHost();
+        return Arrays.stream(currentBundle.getBundleContext().getBundles())
+                .map(b -> new AbstractMap.SimpleEntry<>(b, b.getEntry(path)))
```

Comment:
Note that this implementation relies on jars have directory entries. I believe all modern tools produce jars with directory entries, and other tools like Spring rely on them being present, so it shouldn't be a huge issue.